src/modules/compliance/src/lib/procedures/EnsureInteractiveUsersHomeDirectoriesAreConfigured.cpp

```diff
@@ -0,0 +1,193 @@
+            continue;
+        }
+        const auto pos = line.find("nologin");
+        if (pos != std::string::npos)
+        {
+            OsConfigLogDebug(context.GetLogHandle(), "Ignoring %s entry: %s", etcShellsPath, line.c_str());
+            continue;
+        }
+
+        validShells.insert(line);
+    }
+    return validShells;
+}
+} // anonymous namespace
+
+AUDIT_FN(EnsureInteractiveUsersHomeDirectoriesAreConfigured)
+{
+    UNUSED(args);
+
+    const auto validShells = ListValidShells(context);
+    if (!validShells.HasValue())
+    {
+        OsConfigLogError(context.GetLogHandle(), "Failed to get valid shells: %s", validShells.Error().message.c_str());
+        return validShells.Error();
+    }
+
+    auto cb = [&validShells, &indicators, &context](const passwd* pwd) -> Result<Status> {
+        const auto shell = string(pwd->pw_shell);
+        const auto it = validShells->find(shell);
+        if (it == validShells->end())
+        {
+            OsConfigLogDebug(context.GetLogHandle(), "User '%s' has shell '%s' not listed in %s", pwd->pw_name, pwd->pw_shell, etcShellsPath);
+            return Status::Compliant;
+        }
+
+        struct stat st;
+        if (stat(pwd->pw_dir, &st) != 0)
+        {
+            int status = errno;
+            if (status == ENOENT)
+            {
+                OsConfigLogDebug(context.GetLogHandle(), "User '%s' has home directory '%s' which does not exist", pwd->pw_name, pwd->pw_dir);
+                return indicators.NonCompliant(std::string("User's '") + pwd->pw_name + "' home directory '" + pwd->pw_dir + "' does not exist");
+            }
+            else
+            {
+                OsConfigLogError(context.GetLogHandle(), "Failed to stat home directory '%s' for user '%s': %s", pwd->pw_dir, pwd->pw_name, strerror(status));
+                return Error(string("Failed to stat home directory: ") + strerror(status), status);
+            }
+        }
+
+        const auto* group = getgrgid(pwd->pw_gid);
+        if (nullptr == group)
+        {
+            OsConfigLogError(context.GetLogHandle(), "Failed to get group for user '%s': %s", pwd->pw_name, strerror(errno));
+            return Error(string("Failed to get group for user: ") + strerror(errno), errno);
+        }
+
+        map<string, string> arguments = {{"filename", pwd->pw_dir}, {"mask", "027"}, {"owner", pwd->pw_name}, {"group", group->gr_name}};
+        indicators.Push("EnsureFilePermissions");
+        auto subResult = AuditEnsureFilePermissions(std::move(arguments), indicators, context);
+        indicators.Pop();
+        if (!subResult.HasValue())
+        {
+            OsConfigLogError(context.GetLogHandle(), "Failed to check permissions for home directory '%s' for user '%s': %s", pwd->pw_dir, pwd->pw_name,
+                subResult.Error().message.c_str());
+            return subResult.Error();
+        }
```

Comment:
Relying on other Audit procedures like this is fragile. If we need to share code between the two then lets have the common part in a helper with a proper interface instead.

```
AuditEnsureFilePermissions  \
                             |-------- EnsurePermissions
AuditEnsureInteractive(...) /
```

Reply:
Done